gno.land/cmd/gnoland/main.go

```diff
@@ -314,3 +309,103 @@ func loadGenesisBalances(path string) []string {
 	}
 	return balances
 }
+
+// listGnoPkgs lists all gno packages in the given root directory.
+func listGnoPkgs(root string) ([]pkg, error) {
+	var pkgs []pkg
+
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			goModPath := filepath.Join(path, "gno.mod")
+			data, err := os.ReadFile(goModPath)
+			if err == nil {
+				gnoMod, err := gnomod.Parse(goModPath, data)
+				if err != nil {
+					return fmt.Errorf("parse: %w", err)
+				}
+				gnoMod.Sanitize()
+				if err := gnoMod.Validate(); err != nil {
+					return fmt.Errorf("validate: %w", err)
+				}
+				pkgs = append(pkgs, pkg{
+					name: gnoMod.Module.Mod.Path,
+					path: path,
+					requires: func() []string {
+						var reqs []string
+						for _, req := range gnoMod.Require {
+							reqs = append(reqs, req.Mod.Path)
+						}
+						return reqs
+					}(),
+				})
+				return fs.SkipDir
+			} else if !os.IsNotExist(err) {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pkgs, nil
+}
+
+// sortPkgs sorts the given packages by their dependencies.
+func sortPkgs(pkgs []pkg) error {
+	visited := make(map[string]bool)
+	onStack := make(map[string]bool)
+	sortedPkgs := make([]pkg, 0, len(pkgs))
+
+	var visit func(pkg pkg) error
+	visit = func(pkg pkg) error {
+		if onStack[pkg.name] {
+			return fmt.Errorf("cycle detected: %s", pkg.name)
+		}
+		if visited[pkg.name] {
+			return nil
+		}
+
+		visited[pkg.name] = true
+		onStack[pkg.name] = true
+
+		// Visit package's dependencies
+		for _, req := range pkg.requires {
+			found := false
+			for _, p := range pkgs {
+				if p.name == req {
+					err := visit(p)
+					if err != nil {
+						return err
+					}
+					found = true
+					break
+				}
```

Comment:
opposite check:

```suggestion
				if p.name != req {
					continue
				}
				
				if err := visit(p); err != nil {
					return err
				}
				
				found = true
				break
```

Reply:
Both seem good to me. It comes down to personal preference.

Reply:
It reduces cyclomatic complexity and makes the code more readable. https://go.dev/doc/effective_go#control-structures

Reply:
Resolved-by: a3e978ad739086056b92d14ddb5f27a66df4e14c